docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+# Better Processors
+
+## Introduction
+
+### Background
+
+Conduit processors are currently pretty basic. We would like to make them more powerful
+to give users more freedom when manipulating in-flight data.
+
+Initially, processors were inspired by single message transforms (SMTs) provided by
+Kafka Connect. We tried to mimic the behavior as well as processor names to bring them
+closer to new Conduit users who had prior experience with Kafka Connect. A couple of
+issues arose because of this. The main issue is that Conduit is working with a single
+OpenCDC record that combines the key and payload, whereas Kafka Connect has separate
+payloads for both. The OpenCDC record additionally provides fields like metadata, position,
+operation, and two fields for the payload (before and after). Conduit also differentiates
+between raw data and structured data. All of this meant we tried to jam a square peg into
+a round hole and ended up with sub-par processors.
+
+This document tries to address these issues holistically and create a better experience
+around processors.
+
+### Goals
+
+- Keep existing processors to preserve backward compatibility.
+  - Old processors should only be deprecated but continue to work.
+- Add the ability to run a processor only on records that match a certain condition.
```

Comment:
was this a requirement from the platform team? 
does it mean that the processor won't process the record at all (according to a given condition)? or that the processor will check the record and discard if it doesn't match the condition

Reply:
No, I came up with it 🙃 I realized that it would be useful to be able to have a condition on any processor, not only the `filter` processor. Notice that the processor implementation wouldn't even be aware of this condition, Conduit itself would evaluate it and decide if the processor should be called or skipped. For us it's not much more work, but for the user it's potentially a very useful feature.

Reply:
oh this sounds really useful! nice idea :D